shell/platform/linux/fl_key_event_plugin.cc

```diff
@@ -56,15 +122,149 @@ FlKeyEventPlugin* fl_key_event_plugin_new(FlBinaryMessenger* messenger,
       messenger, channel_name == nullptr ? kChannelName : channel_name,
       FL_MESSAGE_CODEC(codec));
   self->response_callback = response_callback;
+  // Add a weak pointer so we know if the text input plugin goes away.
+  g_object_add_weak_pointer(
+      G_OBJECT(text_input_plugin),
+      reinterpret_cast<gpointer*>(&(self->text_input_plugin)));
+  self->text_input_plugin = text_input_plugin;
 
+  self->pending_events = g_ptr_array_new_with_free_func(g_object_unref);
   return self;
 }
 
-void fl_key_event_plugin_send_key_event(FlKeyEventPlugin* self,
+uint64_t fl_get_event_id(GdkEventKey* event) {
+  // Combine the event timestamp, the type of event, and the hardware keycode
+  // (scan code) of the event to come up with a unique id for this event that
+  // can be derived solely from the event data itself, so that we can identify
+  // whether or not we have seen this event already.
+  return (event->time & 0xffffffff) |
+         (static_cast<uint64_t>(event->type) & 0xffff) << 32 |
+         (static_cast<uint64_t>(event->hardware_keycode) & 0xffff) << 48;
+}
+
+GdkEventKey* fl_find_pending_event(FlKeyEventPlugin* self, uint64_t id) {
+  if (self->pending_events->len == 0 ||
+      FL_KEY_EVENT_PAIR(g_ptr_array_index(self->pending_events, 0))->id != id) {
+    return nullptr;
+  }
+
+  return FL_KEY_EVENT_PAIR(g_ptr_array_index(self->pending_events, 0))->event;
+}
+
+void fl_remove_pending_event(FlKeyEventPlugin* self, uint64_t id) {
+  if (self->pending_events->len == 0 ||
+      FL_KEY_EVENT_PAIR(g_ptr_array_index(self->pending_events, 0))->id != id) {
+    g_warning(
+        "Tried to remove pending event with id %ld, but the event was out of "
+        "order, or is unknown.",
+        id);
+    return;
+  }
+  g_ptr_array_remove_index(self->pending_events, 0);
+}
+
+void fl_add_pending_event(FlKeyEventPlugin* self,
+                          uint64_t id,
+                          GdkEventKey* event) {
+  if (self->pending_events->len > kMaxPendingEvents) {
+    g_warning(
+        "There are %d keyboard events that have not yet received a "
+        "response from the framework. Are responses being sent?",
+        self->pending_events->len);
+  }
+  g_ptr_array_add(self->pending_events, fl_key_event_pair_new(id, event));
+}
+
+struct _KeyEventResponseData {
```

Comment:
Should this be converted into a GObject too?

Reply:
Nevermind: I just went ahead and did it.  It made sense, given the weak pointer to the plugin and all.